src/analysis/monotone-analyzer-impl.h

```diff
@@ -0,0 +1,126 @@
+#ifndef wasm_analysis_monotone_analyzer_impl_h
+#define wasm_analysis_monotone_analyzer_impl_h
+
+#include <iostream>
+#include <unordered_map>
+
+#include "monotone-analyzer.h"
+
+namespace wasm::analysis {
+template<size_t N>
+BlockState<N>::BlockState(const BasicBlock* underlyingBlock)
+  : index(underlyingBlock->getIndex()), cfgBlock(underlyingBlock),
+    states(underlyingBlock->size() + 1, BitsetPowersetLattice<N>::getBottom()),
+    currIndex(0) {}
+
+template<size_t N> void BlockState<N>::addPredecessor(BlockState* pred) {
+  predecessors.push_back(pred);
+}
+
+template<size_t N> void BlockState<N>::addSuccessor(BlockState* succ) {
+  successors.push_back(succ);
+}
+
+template<size_t N> BitsetPowersetLattice<N>& BlockState<N>::getFirstState() {
+  return states.front();
+}
+
+template<size_t N> BitsetPowersetLattice<N>& BlockState<N>::getLastState() {
+  return states.back();
+}
+
+template<size_t N> void BlockState<N>::visitLocalSet(LocalSet* curr) {
+  states[currIndex].value[curr->index] = true;
+}
+
+template<size_t N> void BlockState<N>::visitLocalGet(LocalGet* curr) {
+  states[currIndex].value[curr->index] = true;
+}
+
+template<size_t N> void BlockState<N>::transfer(std::queue<Index>& worklist) {
```

Comment:
It would be nice to have the entire worklist algorithm as a single method of `MonotoneCFGAnalyzer` so the `transfer` function for each block can just take update from an initial state to an end state for a block without worrying about the worklist or any other details of how the analysis is run.

Reply:
I'm leaving this for a follow-up PR. However, the step for propagating states between CFG blocks is now in the `MonotoneCFGAnalyzer`.